server/build_event_protocol/build_event_handler/build_event_handler.go

```diff
@@ -105,25 +106,38 @@ func NewBuildEventHandler(env environment.Env) *BuildEventHandler {
 
 	statsRecorder.Start()
 	webhookNotifier.Start()
+
+	h := &BuildEventHandler{
+		env:              env,
+		statsRecorder:    statsRecorder,
+		openChannels:     openChannels,
+		cancelFnsByInvID: sync.Map{},
+	}
 	env.GetHealthChecker().RegisterShutdownFunction(func(ctx context.Context) error {
+		h.Stop()
 		statsRecorder.Stop()
 		webhookNotifier.Stop()
 		return nil
 	})
-
-	return &BuildEventHandler{
-		env:           env,
-		statsRecorder: statsRecorder,
-		openChannels:  openChannels,
-	}
+	return h
 }
 
 func (b *BuildEventHandler) OpenChannel(ctx context.Context, iid string) interfaces.BuildEventChannel {
 	buildEventAccumulator := accumulator.NewBEValues(iid)
+	ctx = log.EnrichContext(ctx, log.InvocationIDKey, iid)
+	val, ok := b.cancelFnsByInvID.Load(iid)
```

Comment:
can the `sync.Map` be from `CancelFunc -> struct{}` ? It seems a little error prone to depend on the invocation ID being unique in this map, since we have invocation retries. Or if funcs can't be used as sync.Map keys then maybe use a UUID separate from the invocation ID.

Reply:
If sync.Map has the invocation id in the map, we will first call the stored cancel Function and then store the new cancel Function.